hibernate-micrometer/src/main/java/org/hibernate/stat/HibernateMetrics.java

```diff
@@ -0,0 +1,369 @@
+							"region",
+							regionName,
+							"result",
+							"hit"
+					);
+					counter(registry,
+							"hibernate.second.level.cache.requests",
+							"The number of cacheable entities/collections not found in the cache and loaded from the database",
+							stats -> stats.getDomainDataRegionStatistics( regionName ).getMissCount(),
+							"region",
+							regionName,
+							"result",
+							"miss"
+					);
+					counter(
+							registry,
+							"hibernate.second.level.cache.puts",
+							"The number of cacheable entities/collections put in the cache",
+							stats -> stats.getDomainDataRegionStatistics( regionName ).getPutCount(),
+							"region",
+							regionName
+					);
+				} );
+
+		// Entity information
+		counter(registry,
+				"hibernate.entities.deletes",
+				"The number of entity deletes",
+				Statistics::getEntityDeleteCount
+		);
+		counter(registry,
+				"hibernate.entities.fetches",
+				"The number of entity fetches",
+				Statistics::getEntityFetchCount
+		);
+		counter(registry,
+				"hibernate.entities.inserts",
+				"The number of entity inserts",
+				Statistics::getEntityInsertCount
+		);
+		counter(registry, "hibernate.entities.loads", "The number of entity loads", Statistics::getEntityLoadCount );
+		counter(registry,
+				"hibernate.entities.updates",
+				"The number of entity updates",
+				Statistics::getEntityUpdateCount
+		);
+
+		// Collections
+		counter(registry,
+				"hibernate.collections.deletes",
+				"The number of collection deletes",
+				Statistics::getCollectionRemoveCount
+		);
+		counter(registry,
+				"hibernate.collections.fetches",
+				"The number of collection fetches",
+				Statistics::getCollectionFetchCount
+		);
+		counter(registry,
+				"hibernate.collections.loads",
+				"The number of collection loads",
+				Statistics::getCollectionLoadCount
+		);
+		counter(registry,
+				"hibernate.collections.recreates",
+				"The number of collections recreated",
+				Statistics::getCollectionRecreateCount
+		);
+		counter(registry,
+				"hibernate.collections.updates",
+				"The number of collection updates",
+				Statistics::getCollectionUpdateCount
+		);
+
+		// Natural Id cache
+		counter(registry,
+				"hibernate.cache.natural.id.requests",
+				"The number of cached naturalId lookups successfully retrieved from cache",
+				Statistics::getNaturalIdCacheHitCount,
+				"result",
+				"hit"
+		);
+		counter(registry,
+				"hibernate.cache.natural.id.requests",
+				"The number of cached naturalId lookups not found in cache",
+				Statistics::getNaturalIdCacheMissCount,
+				"result",
+				"miss"
+		);
+		counter(registry, "hibernate.cache.natural.id.puts", "The number of cacheable naturalId lookups put in cache",
+				Statistics::getNaturalIdCachePutCount
+		);
+
+		counter(registry,
+				"hibernate.query.natural.id.executions",
+				"The number of naturalId queries executed against the database",
+				Statistics::getNaturalIdQueryExecutionCount
+		);
+
+		TimeGauge.builder(
```

Comment:
I wonder if we should wrap this in a `TimeWindowMax` so this value decays more usefully. Could also publish both this "all-time" max and a decaying one separately.

Reply:
@sebersole ? This is one of those questions that comes back to which metrics are useful. ;)

I'm tempted to leave this one as it is/was in 5.4 (5.x?)

Reply:
Well I think this comes back to the generalized approach we want here...  The general consensus seemed to be:

1) keep exposed metrics as they are today in your existing integration for 5.x
2) redesign/redefine the metrics that actually make sense to expose for 6.x

I'm not sure that second part will ever come to fruition - people become loathe to change things after they are exposed.  Even more so here I would think for a "soft API".  But y'all have more experience with changing these "APIs" from release to release, so I'll defer to what you think.

Reply:
In a way, it's an optionally used thing.. people can always copy the existing one if they need to keep existing behavior... which is why I can see an argument for updating the behavior here if that makes more sense.

OTOH, this is also an older stream, so I could justify doing more along the lines of "recommended metrics" in version 6 (as there are a lot of other changes in the area, too).

What isn't apparent from the 5.x API is which metrics are there for historical reasons, which are there for sanity checking / debugging, and which are actually useful for gaining some notion of health or performance you'd want to monitor periodically over time, etc.

So I can go either way here: pick up TimeWindowMax so the max decays usefully (or publish both to preserve existing values), or leave everything as-is/was, and we can make better recommendations in the 6.x stream.

Reply:
> In a way, it's an optionally used thing, people can always copy the existing one if they need to keep existing behavior. Which is why I can see either side as valid. ;)

Unless I misunderstand, simply dropping in the old one would not work as I'd plan on making at least some of the corresponding changes to the Statistics API in 6.  Many of those do not make much sense, and the cache region name stuff at the very least need to be simplified


> What isn't apparent from the 5.x API is which metrics are there for historical reasons, which are there for sanity checking / debugging, and which are actually useful for gaining some notion of health or performance you'd want to monitor periodically over time, etc.

I was not involved in the original metrics impl nor consumed it, so I'm not the one to answer that.


> So whichever you want: if you think including better decaying behavior in 5.x is a cool thing to do, I can add it. (both existing max + decaying max?) If you would rather focus improvement on gathered metrics for 6.x, I'm fine w/ that, too.

I think the decaying max is more useful.  And I just can't imagine that the distinction is going to negatively impact anyone.  Famous last words...

Reply:
I'll add it to the 5.x one then. 6.x has API changes.. and I think we can/should redo that one to contain useful metrics, full stop. ;) If that means dropping things, it means dropping things (but 5 -> 6 should allow that)